server/src/main/java/org/opensearch/search/internal/SearchContext.java

```diff
@@ -188,6 +188,10 @@ public final void close() {
 
     public abstract void highlight(SearchHighlightContext highlight);
 
+    public boolean hasInnerHits() {
```

Comment:
@jainankitk I don't understand what this pull request is doing to be fair (apologies for that): to me, it replaces the `null` checks with `hasXxx` method (which conceptually is cleaner) but it does not remove any unused sub phase initialization logic (instead introduce weird dependency between  `hasInnerHits()`  and `innerHits()` methods, whichever is called first, which does not look right)

Reply:
@reta - Thats a fair question, and I am glad you asked. During the node initialization, possible fetch sub phase handlers are registered [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/SearchModule.java#L1047) and during fetch phase execution, the fetch sub processors for the ones relevant to search request get initialized. While debugging code, I noticed that the scriptFields and innerHits subprocessors are always initialized irrespective of the search request. For example - below request does not have anything to do with either:

```
curl -XGET "http://localhost:9200/logs-241998/_search?pretty" -H 'Content-Type: application/json' -d'
{"size": 1000,
  "query": {
    "match_all": {}
  },"docvalue_fields":["clientip"],
 "sort": [{"size":"asc"}]
}'
```
Then, I noticed that implementation of `scriptFields()` and `innerHits()` always returns non-null values:

```
public InnerHitsContext innerHits() {
        if (innerHitsContext == null) {
            innerHitsContext = new InnerHitsContext();
        }
        return innerHitsContext;
    }
```

so existing checks were useless:

```
if (fetchContext.innerHits() == null) //It is always non-null!!
```

Hence, I went ahead and fixed the logic by introducing another `hasInnerHits()` method to not change anything else with original and regress elsewhere.

Reply:
Thanks @jainankitk , so why we need to initialize in first place if the code seems to deal with `null` properly?

```
public InnerHitsContext innerHits() {
        return innerHitsContext;
    }

```

Reply:
Because all the other references of `innerHits()` & `scriptFields()` assume implicit initialization. This was the only place where the method was used for checking the initialization which is of course incorrect as output is guaranteed to be non-null. For example - below invocation assumed scriptFields is non-null:

```
for (ScriptFieldsContext.ScriptField field : scriptFields) {
            subSearchContext.scriptFields().add(field);
        }
```

Similarly, for innerHits:

```
if (innerHitBuilders.size() > 0) {
            for (Map.Entry<String, InnerHitContextBuilder> entry : innerHitBuilders.entrySet()) {
                try {
                    entry.getValue().build(context, context.innerHits());
                } catch (IOException e) {
                    throw new SearchException(shardTarget, "failed to build inner_hits", e);
                }
            }
        }
```

Reply:
Thanks @jainankitk , fair enough, so here the issue I have with this change:  calling `innerHits()` and `hasInnerHits()` != calling `hasInnerHits()` and `innerHits()`, this is for obvious reasons not a good design decision. I would suggest to explore the routes to:
a) let `innerHits()` return `Optional<InnerHitsContext>`
b) always initialize the `innerHitsContext`

Reply:
Sorry I am not exactly clear about your concern here. We have similar `hasScriptFields` for `scriptFields` already, it was not getting used. Regarding the approaches, I am not completely sure what you're suggesting:

> always initialize the innerHitsContext

Why should we do that if innerHitsContext is not needed for many cases?

> let innerHits() return Optional<InnerHitsContext>

Since I am not sure how it will work, maybe you can take a quick first stab at it. Also, to make it consistent with other attributes like `scriptFields`, we need to change others as well. Does it make sense to make that change separately?

Reply:
@jainankitk you are very right :((( we sadly have this pattern all over the place, it is not worth addressing one place :(